File: leetcode/Python/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        st, d = [], {'(':')', '{':'}', '[':']'}
        for v in s:
            if v in d.keys():
                st.append(d[v])
            elif st and v==st[-1]:
                st.pop()
            else:
                return False
        return not st


'''
class Solution:
    def isValid(self, s: str) -> bool:
        um, st = {'(': ')', '{': '}', '[':']'}, list()
        for ch in s:
            t = um.get(ch, None)
            if t is not None:
                st.append(t)
            elif len(st)>0 and st[-1] == ch:
                st.pop()
            else:
                return False
        return len(st) == 0
'''

File: leetcode/C++/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        unordered_map<char,char> um={{'(',')'},{'[',']'},{'{','}'}};
        for(char &t : s){
            if(t=='('||t=='{'||t=='[') st.push(t);
            else if(!st.empty() && um[st.top()]==t) st.pop();
            else return false;
        }
        return st.empty();
    }
};

/*
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char,char> um = {{'(', ')'}, {'[', ']'}, {'{', '}'}};
        unordered_map<char,char>::iterator it;
        vector<char> st;
        for(char &ch : s){
            if((it=um.find(ch))!=um.end()) st.push_back(it->second);
            else if(!st.empty() && st.back()==ch) st.pop_back();
            else return false;
        }
        return st.empty();
    }
};
*/